File: src/main/java/com/github/sorusclient/client/module/ModuleManager.java
package com.github.sorusclient.client.module;

import com.github.sorusclient.client.Sorus;
import com.github.sorusclient.client.module.impl.blockoverlay.BlockOverlay;
import com.github.sorusclient.client.module.impl.fullbright.FullBright;
import com.github.sorusclient.client.module.impl.itemphysics.ItemPhysics;
import com.github.sorusclient.client.module.impl.oldanimations.OldAnimations;
import com.github.sorusclient.client.module.impl.environmentchanger.EnvironmentChanger;
import com.github.sorusclient.client.module.impl.perspective.Perspective;
import com.github.sorusclient.client.module.impl.zoom.Zoom;
import com.github.sorusclient.client.setting.SettingManager;
import com.github.sorusclient.client.util.Pair;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class ModuleManager {

    private final Map<Class<Module>, ModuleData> modules = new HashMap<>();

    public void initialize() {
        this.registerInternalModules();
    }

    private void registerInternalModules() {
        this.register(new BlockOverlay(), "BlockOverlay", "test");
        this.register(new FullBright(), "FullBright", "test");
        this.register(new ItemPhysics(), "ItemPhysics", "test");
        this.register(new OldAnimations(), "OldAnimations", "test");
        this.register(new Perspective(), "Perspective", "test");
        this.register(new EnvironmentChanger(), "EnvironmentChanger", "test");
        this.register(new Zoom(), "Zoom", "test");
    }

    @SuppressWarnings("unchecked")
    public void register(Module module, String name, String description) {
        this.modules.put((Class<Module>) module.getClass(), new ModuleData(module, name, description));
        Sorus.getInstance().get(SettingManager.class).register(module);
    }

    @SuppressWarnings("unchecked")
    public <T> T get(Class<T> moduleClass) {
        return (T) this.modules.get(moduleClass).getModule();
    }

    public List<ModuleData> getModules() {
        return new ArrayList<>(this.modules.values());
    }

}

File: src/main/kotlin/com/github/sorusclient/client/module/ModuleManager.kt
package com.github.sorusclient.client.module

import com.github.sorusclient.client.setting.SettingManager

object ModuleManager {
    val modules: MutableMap<Class<out Module>, ModuleData> = HashMap()

    fun initialize() {
        registerInternalModules()
    }

    private fun registerInternalModules() {
    }

    fun register(module: Module, name: String, description: String) {
        modules[module.javaClass] =
            ModuleData(module, name, description)
        SettingManager.register(module)
    }

    inline fun <reified T: Module> get(): T {
        return modules[T::class.java]!!.module as T
    }

    operator fun get(id: String): Module? {
        val optionalModuleData = modules.values.stream().filter { moduleData: ModuleData -> moduleData.module.id == id }
            .findFirst()
        return optionalModuleData.map { obj: ModuleData -> obj.module }.orElse(null)
    }

}